downloader/src/main/kotlin/VersionControlSystem.kt

```diff
@@ -283,11 +283,17 @@ abstract class VersionControlSystem {
             throw DownloadException("Unable to determine a revision to checkout.")
         }
 
-        var i = 0
-        val workingTreeRevision = revisionCandidates.find { revision ->
-            log.info { "Trying revision candidate '$revision' (${++i} of ${revisionCandidates.size})..." }
-            updateWorkingTree(workingTree, revision, pkg.vcsProcessed.path, recursive)
-        } ?: throw DownloadException("$type failed to download from URL '${pkg.vcsProcessed.url}'.")
+        val results = mutableListOf<Result<String>>()
+
+        revisionCandidates.forEachIndexed { index, revision ->
+            log.info { "Trying revision candidate '$revision' (${index + 1} of ${revisionCandidates.size})..." }
+            results += updateWorkingTree(workingTree, revision, pkg.vcsProcessed.path, recursive)
+            if (results.last().isSuccess) return@forEachIndexed
+        }
+
+        val workingTreeRevision = results.last().getOrElse {
+            throw DownloadException("$type failed to download from URL '${pkg.vcsProcessed.url}'.", it)
```

Comment:
In the `if (revisionCandidates.isEmpty())` case we still don't propagate the exceptions that could have happened before (lines 233 and 253). I think it would be good to add them to the exception thrown in line 283 to completely address #3952 . Also I wonder if it is sufficient to figure out the problem to only propagate the last exception. But it probably is, at least I have no idea currently in which situation it wouldn't be.

Reply:
Issue #3952 specifically talks about `updateWorkingTree`, so that's what I was looking at, and the current changes already properly provide the "too long path" root cause I'm looking at in #4024. But let's see what I can do about the other exceptions...

Reply:
Thanks.